### PR TITLE
Adding Table.removeSearchIndex() to remove search index, and renaming other index methods

### PR DESCRIPTION
**Pull request by kneth**
No updates to change log since it's an internal method.

@emanuelez @cmelchior

### CODE REVIEW

**Comment by cmelchior** [comment]
Why unset instead of remove which better fits the naming we use in the rest of API?

**Comment by kneth** [comment]
Good question. But `set` and `remove` don't sounds right. Maybe `add` and `remove` do - or `set` and `unset` :-)

**Comment by cmelchior** [comment]
We use set for PrimaryKey but that is because there can only be one. I like add and remove more for Index as you can have multiple

**Comment by kneth** [comment]
I propose that that we have the names to `addSearchIndex`, `removeSearchIndex`, and `hasSearchIndex`.

**Comment by bmunkholm** [comment]
I propose you align with Cocoa - that should be the default to do unless there are very specific platform standards to follow.

**Comment by kneth** [comment]
The Cocoa binding doesn't have public method for it. The Java binding need to wrap the core method but we don't expose them directly and only through the `io.realm.internal` package.

File: realm/src/androidTest/java/io/realm/internal/JNITransactions.java [hunk 1]
```diff
@@ -255,7 +255,7 @@ public void testMustFailOnWriteInReadTransactions() {
         try { table.setBinaryByteArray(0,0,null);   fail();} catch (IllegalStateException e) {assertNotNull(e);}
         try { table.setBoolean(0,0,false);          fail();} catch (IllegalStateException e) {assertNotNull(e);}
         try { table.setDate(0,0,new Date(0));       fail();} catch (IllegalStateException e) {assertNotNull(e);}
-        try { table.setIndex(0);                    fail();} catch (IllegalStateException e) {assertNotNull(e);}
+        try { table.addSearchIndex(0);                    fail();} catch (IllegalStateException e) {assertNotNull(e);}
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/internal/JNITransactions.java]: indent...

File: realm/src/androidTest/java/io/realm/internal/TableIndexAndDistinctTest.java [hunk 2]
```diff
@@ -148,5 +141,23 @@ public void testShouldThrowDistinctErrorWhenWrongColumnType() {
         List<?> mixedValues = Arrays.asList(values);
         return DataProviderUtil.allCombinations(mixedValues);
     }*/
+
+    public void testUnsetIndex() {
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/internal/TableIndexAndDistinctTest.java]: naming wrong

File: realm/src/androidTest/java/io/realm/internal/TableIndexAndDistinctTest.java [hunk 3]
```diff
@@ -148,5 +141,23 @@ public void testShouldThrowDistinctErrorWhenWrongColumnType() {
...
+        init();
+        table.addSearchIndex(1);
+        assertEquals(true, table.hasSearchIndex(1));
+
+        table.removeSearchIndex(1);
+        assertEquals(false, table.hasSearchIndex(1));
+    }
+
+    public void testUnsetIndexNoop() {
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/internal/TableIndexAndDistinctTest.java]: naming wrong

File: realm-jni/src/io_realm_internal_table.cpp [hunk 4]
```diff
@@ -740,6 +740,22 @@ JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeSetIndex(
     } CATCH_STD()
 }
 
+JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeUnsetIndex(
+    JNIEnv* env, jobject, jlong nativeTablePtr, jlong columnIndex)
+{
+    Table* pTable = TBL(nativeTablePtr);
+    if (!TBL_AND_COL_INDEX_VALID(env, pTable, columnIndex))
+        return;
+    if (pTable->get_column_type (S(columnIndex)) != type_String) {
```
Comment (bmunkholm) [inline realm-jni/src/io_realm_internal_table.cpp]: Should that still be the limit now? (or is it fixed in another PR)?

File: realm-jni/src/io_realm_internal_table.cpp [hunk 5]
```diff
@@ -740,6 +740,22 @@ JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeSetIndex(
     } CATCH_STD()
 }
 
+JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeUnsetIndex(
```
Comment (bmunkholm) [inline realm-jni/src/io_realm_internal_table.cpp]: Rename to match Java

File: realm/src/main/java/io/realm/internal/Table.java [hunk 6]
```diff
@@ -1152,15 +1152,24 @@ public void adjust(long columnIndex, long value) {
...
     protected native void nativeAddInt(long nativeViewPtr, long columnIndex, long value);
 
+    private final String MSG_ONLY_STRING_COLOUM_CAN_BE_INDEXED = "Index is only supported on string columns.";
 
-    public void setIndex(long columnIndex) {
+    public void addSearchIndex(long columnIndex) {
         checkImmutable();
         if (getColumnType(columnIndex) != ColumnType.STRING) {
-            throw new IllegalArgumentException("Index is only supported on string columns.");
+            throw new IllegalArgumentException(MSG_ONLY_STRING_COLOUM_CAN_BE_INDEXED);
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/internal/Table.java]: Remove these checks on Java side. Place them in JNI C++ layer.

File: realm/src/main/java/io/realm/internal/Table.java [hunk 7]
```diff
@@ -1152,15 +1152,24 @@ public void adjust(long columnIndex, long value) {
...
         }
         nativeSetIndex(nativePtr, columnIndex);
     }
 
+    public void removeSearchIndex(long columnIndex) {
+        checkImmutable();
+        if (getColumnType(columnIndex) != ColumnType.STRING) {
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/internal/Table.java]: Remove redundant check - you also have it in JNI C++ layer.

File: realm/src/main/java/io/realm/internal/Table.java [hunk 8]
```diff
@@ -1211,8 +1220,9 @@ Group getTableGroup() {
 
     protected native void nativeSetIndex(long nativePtr, long columnIndex);
 
+    protected native void nativeUnsetIndex(long nativePtr, long columnIndex);
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/internal/Table.java]: rename to match Java naming.

File: realm-jni/src/io_realm_internal_table.cpp (hunk 4 above)
Comment (kneth) [inline realm-jni/src/io_realm_internal_table.cpp]: I rather do it a separate PR as it will require many changes and additional tests.

File: realm/src/androidTest/java/io/realm/internal/JNITransactions.java (hunk 1 above)
Comment (kneth) [inline realm/src/androidTest/java/io/realm/internal/JNITransactions.java]: Android Studio's refactoring isn't that smart :-(

**Comment by kneth** [comment]
retest please

**Comment by kneth** [comment]
retest this please

**Comment by kneth** [comment]
retest this please

**Comment by kneth** [comment]
@cmelchior @emanuelez Please review.

**Comment by cmelchior** [comment]
:+1:

**Comment by bmunkholm** [comment]
:+1:
